### PR TITLE
[CALCITE-7634] JoinExpandOrToUnionRule incorrectly expands OR branches with non-equi predicates referencing both join inputs

### PR DESCRIPTION
**Pull request by xiedeyantu**
## Jira Link

[CALCITE-7634](https://issues.apache.org/jira/browse/CALCITE-7634)

### CODE REVIEW

File: core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java [hunk 1]
```diff
@@ -10008,6 +10008,19 @@ private RelOptFixture spatial(String sql) {
         .check();
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-7634">[CALCITE-7634]
+   * JoinExpandOrToUnionRule incorrectly expands OR branches with non-equi
+   * predicates referencing both join inputs</a>. */
+  @Test void testJoinConditionOrExpansionRuleWithCrossInputPredicate() {
```
Comment (xuzifu666) [inline core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java:10015]: This only asserts the plan shape, maybe it can't catch the row-dropping bug above. Can we add a quidem test for it?  that runs with NULL data and checks the actual result set.

**Review by xuzifu666 — APPROVED** [review]
Overall it is a nice fix, only left a comment for test.

**Comment by mihaibudiu** [comment]
We should merge correctness bugfixes as soon as possible.

**Comment by xiedeyantu** [comment]
I added test in hep.iq. I changed the NOT to IS NOT TRUE. Please take another look. Thanks! @xuzifu666 @mihaibudiu

**Comment by xuzifu666** [comment]
Thanks for the test; I think everything looks good now. However, the Jira ticket doesn't seem to have a link to this PR yet—you might need to add it manually. @xiedeyantu

**Comment by xiedeyantu** [comment]
If there are no other comments, I plan to merge them in a day.
